### ISSUE TITLE
StringTie not creating output dir automatically

### ISSUE THREAD

**Issue by lcolladotor** [gpertea/stringtie#30]
Hi,

I ran StringTie interactively (qrsh) on the JHPCE cluster and it would create the output directory but hang at that point. When I tried running StringTie via submitted jobs (qsub) the job would fail because the output directory was not created by StringTie. From the docs, I understand that StringTie does this step. Anyhow, simply creating the output directory before running StringTie solves the problem. 

So, if you just don't want to support this small issue, simply update the manual.

Here is an example of the type of error I was getting (via qsub):

```
Error creating output file
/dcs01/ajaffe/Brain/derRuns/derSupplement/simulation/stringtie/sample1G1R1/outfile.gtf.tmp
```

where `/dcs01/ajaffe/Brain/derRuns/derSupplement/simulation/stringtie` existed already, but not the subdirectory `sample1G1R1`.

This is with StringTie version 1.1.2.

Best,
Leo

**Comment by gpertea** [comment]
StringTie is indeed supposed to create the output directory, as described in the manual, and in my tests it does actually do it, so I suspect there is something fishy with the way I call mkdir() there which does not quite work as expected on a grid file system (session) like the one you are using. Ah, I see that I attempt to create the new directory with permissions 0775 and I neglected to check if the mkdir() call was actually successful -- this is definitely something I should fix, thank you for bringing it to my attention!

Just out of curiosity, what is the command line you used (what options) ?

**Comment by lcolladotor** [comment]
The script I used is [this one](https://github.com/lcolladotor/derSupplement/blob/gh-pages/simulation/stringtie/run-stringtie.sh). The error popped up during the first StringTie run. That is:

`stringtie ${fullpath} -o ${WDIR}/${libname}/outfile.gtf -p ${CORES} -G ${WDIR}/hg19-knownGene.GTF`

**Comment by lcolladotor** [comment]
Regarding the qrsh attempt, I was wrong. It does not hang, I simply
didn't wait enough for the job to complete. I realized this after
using the -v option and completed the test with a smaller GTF file
that the one I was using before.

**Comment by gpertea** [comment]
So through qrsh  it actually created the output directory by itself, as expected? 
Anyway, I've already made some adjustments to the mkdir() code, now it should be more explicit about any system errors and it's more conservative about the permissions requested for the created directory (I suspect it was a permission issue with those submitted jobs which failed for you). These adjustments should be in the next release..

**Comment by lcolladotor** [comment]
Yes, on my most recent test qrsh actually created the output directory as expected. So the problem was only with qsub.

Thank you for addressing this small issue =)

**Comment by gpertea** [comment]
Could you please try v1.2.0 of StringTie (we just released) on a small example and let me know if the new mkdir() code now works as it should with qsub ?

**Comment by lcolladotor** [comment]
Hi,

Using version 1.2.1 I get the following error:

```
Warning: failed at mkdir(): No such file or directory
Error: cannot create directory /dcs01/ajaffe/Brain/derRuns/derSupplement/simulation/stringtie/sample1G1R1-no-assembly/!
```

That comes from using https://github.com/lcolladotor/derSupplement/blob/af6b329366181a978c3845843b846ffc0a2660da/simulation/stringtie/run-stringtie.sh#L198 which looks like this:

```bash
stringtie ${fullpath} -o ${WDIR}/${libname}-no-assembly/outfile.gtf -p ${CORES} -G ${MAINDIR}/gtf/chr17.gtf -b ${WDIR}/${libname}-no-assembly/ -e
```

The problem goes away by creating the directory manually before running stringtie (see https://github.com/lcolladotor/derSupplement/commit/d47d45fc220db62833b5ff5a0e95efc1759c0376).

I just removed the redudant `-b` info like you've told me before with commit https://github.com/lcolladotor/derSupplement/commit/5a758d7f0a887e34d388c6d638b8f55edf5f7c2c

Sorry for not checking back in a while.

Best,
Leo

**Comment by gpertea** [comment]
Weird, not sure why mkdir() still fails under qsub, I have to take another look.. 
For now I can only comment related to the `-b <output_dir>` option,  it seems you want use the `-B` option (instead of `-b`) if you still want the ballgown tables, and they'll be written in the same output directory as the one specified by the `-o` option.

**Comment by lcolladotor** [comment]
Yup, thanks! I noticed the `-b` vs `-B` issue as well (https://github.com/lcolladotor/derSupplement/commit/253f05415734d1f0e60632b476c0a94949efc1ed).

**Comment by FredericLammer** [comment]
Hello gpertea,

I'd first like to thank you for providing such an amazingly fast and versatile RNA seq analysis tool. After upgrading to stringtie v1.2.0 / v1.2.1 I'm constantly getting error messages and empty output tables/no output tables at all. Neither the-b command nor the combination of -B and -o seem to work, it's always a  

>Error creating output file:

message. The error message that pops up when using the -b option provides a temporary directory while the -B/-o combination provides the folder specified by -o. I first assumed that the error might have something to do with the newly added --merge command in stringtie v1.2.0 but sadly, running the -b (or the -B/-o) option with a cuffmerge gtf-file results in the same error message. 

On an additional note, I've never seen any temporary directory at all when using older versions of stringtie, but I don't know if this might help you resolving the (presumably) buggy ballgown output feature.

Regards,
Frederic

**Comment by gpertea** [comment]
Wow this sounds like a more common issue than I thought! 
Indeed since v1.2.0 I started using *mkdtemp()* to create the temporary directory and I have changed slightly the way the output files are managed - of course I have not encountered any problems with these changes on my system(s) (mostly RedHat based, on local or NFS shares, and some older OS X system). 
      Frederic, could you please provide me with more details about the system(s) where you are getting these errors? Also I'd like to know about the underlying filesystem - is it a network share of some sort, perhaps coming from a Windows server? (or whatever it is, it would be good to know).
     If you prefer you can send this info directly to gpertea@jhu.edu directly instead if posting it here.

**Comment by gpertea** [comment]
It would be great if I could get some feedback on this issue with the new v1.2.2 version. I took another shot at the automatic path creation and I'd like to see if anyone still encounters any directory creation errors, and if so, please report here (or directly to me) the exact error message.

**Comment by sum732** [comment]
I am running stringtie-1.3.3b.Linux_x86_64 version.
Command used: /Programs/StringTie/stringtie-1.3.3b.Linux_x86_64/stringtie ../star/Sample_212a.Aligned.sortedByCoord.out.bam -o Sample_212a -p 5 -G /data/Human/GRCh37/gencode.v19.annotation.A.R.1.gtf --fr -f 0.01 -A gene -C -B -v
Error: Error creating output file Sample_212a
Does not matter, if I run it locally or on a cluster. StingTie keeps giving the same error.

**Comment by golrokh51** [comment]
Same problem with stringtie/2.1.0 on HPC
`slurmstepd: error: Unable to create TMPDIR [/localscratch/golrokh.48950538.0]: Permission denied`
`slurmstepd: error: Setting TMPDIR to /tmp`
However, when I am connected directly to a calculating node interactively, there is no error
